From 03d01b417930aa6d44e76d55c51f928ab347108a Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@nxp.com>
Date: Fri, 15 Jun 2018 14:09:33 +0800
Subject: [PATCH 4141/5242] MLK-18609-3 usb: cdns3: gadget: refine the memory
 allocation function

commit  f4e05469f694eb131dc739d840d619071dfcd168 from
https://source.codeaurora.org/external/imx/linux-imx.git

Move all memory allocation to usb_ss_allocate_trb_pool and align with
usb_ss_free_trb_pool.

Reviewed-by: Jun Li <jun.li@nxp.com>
Signed-off-by: Peter Chen <peter.chen@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/cdns3/gadget.c |   28 +++++++++++-----------------
 1 file changed, 11 insertions(+), 17 deletions(-)

diff --git a/drivers/usb/cdns3/gadget.c b/drivers/usb/cdns3/gadget.c
index 57b37a1..3b007ce 100644
--- a/drivers/usb/cdns3/gadget.c
+++ b/drivers/usb/cdns3/gadget.c
@@ -197,18 +197,20 @@ static int usb_ss_allocate_trb_pool(struct usb_ss_endpoint *usb_ss_ep)
 {
 	struct usb_ss_dev *usb_ss = usb_ss_ep->usb_ss;
 
-	if (usb_ss_ep->trb_pool)
-		return 0;
-
-	usb_ss_ep->trb_pool = dma_zalloc_coherent(usb_ss->sysdev,
+	if (!usb_ss_ep->trb_pool) {
+		usb_ss_ep->trb_pool = dma_zalloc_coherent(usb_ss->sysdev,
 			sizeof(struct usb_ss_trb) * USB_SS_TRBS_NUM,
 		&usb_ss_ep->trb_pool_dma, GFP_DMA);
+		if (!usb_ss_ep->trb_pool)
+			return -ENOMEM;
+	}
 
-	if (!usb_ss_ep->trb_pool) {
-		dev_err(&usb_ss->dev,
-				"Failed to allocate TRB pool for endpoint %s\n",
-				usb_ss_ep->name);
-		return -ENOMEM;
+	if (!usb_ss_ep->cpu_addr) {
+		usb_ss_ep->cpu_addr = dma_alloc_coherent(usb_ss->sysdev,
+				CDNS3_UNALIGNED_BUF_SIZE,
+				&usb_ss_ep->dma_addr, GFP_DMA);
+		if (!usb_ss_ep->cpu_addr)
+			return -ENOMEM;
 	}
 
 	return 0;
@@ -1548,14 +1550,6 @@ static int usb_ss_gadget_ep_enable(struct usb_ep *ep,
 	if (ret)
 		return ret;
 
-	if (!usb_ss_ep->cpu_addr) {
-		usb_ss_ep->cpu_addr = dma_alloc_coherent(usb_ss->sysdev,
-				CDNS3_UNALIGNED_BUF_SIZE,
-				&usb_ss_ep->dma_addr, GFP_DMA);
-		if (!usb_ss_ep->cpu_addr)
-			return -ENOMEM;
-	}
-
 	dev_dbg(&usb_ss->dev, "Enabling endpoint: %s, addr=0x%x\n",
 		ep->name, desc->bEndpointAddress);
 	spin_lock_irqsave(&usb_ss->lock, flags);
-- 
1.7.9.5

